File: backend/etl/utils/ppc_helper.py
from datetime import date
from sqlalchemy.orm import Session
from database.models import PpcSequence, AdvanceRequest, Settlement
from sqlalchemy import func

# MONTH TO ROMAN
def month_to_roman(month: int) -> str:
    roman_months = {
        1: "I",
        2: "II",
        3: "III",
        4: "IV",
        5: "V",
        6: "VI",
        7: "VII",
        8: "VIII",
        9: "IX",
        10: "X",
        11: "XI",
        12: "XII",
    }

    return roman_months[month]

# GET NEXT SEQUENCE
def get_next_ppc_sequence(db: Session, year: int) -> int:
    # 1. Query max sequence dari PpcSequence (audit log)
    seq_from_history = (
        db.query(func.max(PpcSequence.sequence))
        .filter(PpcSequence.year == year)
        .scalar()
    ) or 0

    # 2. Safety fallback: Query max sequence dari AdvanceRequest
    existing_advances = (
        db.query(AdvanceRequest.ppc_no)
        .filter(AdvanceRequest.ppc_no.like(f"%/{year}"))
        .all()
    )
    max_adv_seq = 0
    for item in existing_advances:
        if item[0]:
            try:
                max_adv_seq = max(max_adv_seq, int(item[0].split("/")[0]))
            except Exception:
                pass

    # 3. Safety fallback: Query max sequence dari Settlement
    existing_settlements = (
        db.query(Settlement.ppc_no)
        .filter(Settlement.ppc_no.like(f"%/{year}"))
        .all()
    )
    max_settle_seq = 0
    for item in existing_settlements:
        if item[0]:
            try:
                max_settle_seq = max(max_settle_seq, int(item[0].split("/")[0]))
            except Exception:
                pass

    return max(seq_from_history, max_adv_seq, max_settle_seq) + 1

# RECORD PPC SEQUENCE
def record_ppc_sequence(db: Session, ppc_no: str, year: int, sequence: int):
    existing = (
        db.query(PpcSequence)
        .filter(PpcSequence.ppc_no == ppc_no)
        .first()
    )
    if not existing:
        seq_record = PpcSequence(
            ppc_no=ppc_no,
            year=year,
            sequence=sequence
        )
        db.add(seq_record)

# GENERATE PPC NUMBER
# FORMAT: {sequence}/PPC/{roman_month}/{year}
def generate_ppc_no(
    db: Session,
    request_date: date,
    record_sequence: bool = False,
) -> str:
    year = request_date.year
    roman_month = month_to_roman(request_date.month)

    next_sequence = get_next_ppc_sequence(db, year)
    ppc_no = f"{next_sequence}/PPC/{roman_month}/{year}"

    if record_sequence:
        record_ppc_sequence(
            db=db,
            ppc_no=ppc_no,
            year=year,
            sequence=next_sequence
        )

    return ppc_no

File: backend/services/PpcHelper.php
<?php
require_once __DIR__ . '/../config/database.php';

class PpcHelper {
    const STARTING_PPC_SEQUENCE = 648;
    const STARTING_PPC_SEQUENCE_YEAR = 2026;

    public static function monthToRoman(int $month): string {
        $romanMonths = [
            1 => "I", 2 => "II", 3 => "III", 4 => "IV",
            5 => "V", 6 => "VI", 7 => "VII", 8 => "VIII",
            9 => "IX", 10 => "X", 11 => "XI", 12 => "XII"
        ];
        return $romanMonths[$month] ?? "I";
    }

    public static function getNextPpcSequence(PDO $db, int $year): int {
        // Query max sequence from ppc_sequences
        $stmtHistory = $db->prepare("SELECT MAX(sequence) AS max_seq FROM ppc_sequences WHERE year = :yr");
        $stmtHistory->execute([':yr' => $year]);
        $rowHistory = $stmtHistory->fetch();
        $seqFromHistory = $rowHistory && $rowHistory['max_seq'] !== null ? (int)$rowHistory['max_seq'] : 0;

        // Safety fallback: query max sequence from advance_requests
        $stmtAdv = $db->prepare("SELECT ppc_no FROM advance_requests WHERE ppc_no LIKE :pattern");
        $stmtAdv->execute([':pattern' => "%/{$year}"]);
        $advances = $stmtAdv->fetchAll();
        $maxAdvSeq = 0;
        foreach ($advances as $item) {
            if (!empty($item['ppc_no'])) {
                $parts = explode('/', $item['ppc_no']);
                if (isset($parts[0]) && is_numeric($parts[0])) {
                    $maxAdvSeq = max($maxAdvSeq, (int)$parts[0]);
                }
            }
        }

        // Safety fallback: query max sequence from settlements
        $stmtSettle = $db->prepare("SELECT ppc_no FROM settlements WHERE ppc_no LIKE :pattern");
        $stmtSettle->execute([':pattern' => "%/{$year}"]);
        $settlements = $stmtSettle->fetchAll();
        $maxSettleSeq = 0;
        foreach ($settlements as $item) {
            if (!empty($item['ppc_no'])) {
                $parts = explode('/', $item['ppc_no']);
                if (isset($parts[0]) && is_numeric($parts[0])) {
                    $maxSettleSeq = max($maxSettleSeq, (int)$parts[0]);
                }
            }
        }

        $currentMax = max($seqFromHistory, $maxAdvSeq, $maxSettleSeq);
        if ($year === self::STARTING_PPC_SEQUENCE_YEAR && $currentMax < self::STARTING_PPC_SEQUENCE) {
            return self::STARTING_PPC_SEQUENCE + 1;
        }
        return $currentMax + 1;
    }

    public static function recordPpcSequence(PDO $db, string $ppcNo, int $year, int $sequence): void {
        $checkStmt = $db->prepare("SELECT id FROM ppc_sequences WHERE ppc_no = :ppc_no");
        $checkStmt->execute([':ppc_no' => $ppcNo]);
        if (!$checkStmt->fetch()) {
            $stmt = $db->prepare("
                INSERT INTO ppc_sequences (ppc_no, year, sequence, issued_at) 
                VALUES (:ppc_no, :yr, :seq, GETDATE())
            ");
            $stmt->execute([
                ':ppc_no' => $ppcNo,
                ':yr' => $year,
                ':seq' => $sequence
            ]);
        }
    }

    public static function generatePpcNo(PDO $db, DateTime $requestDate, bool $recordSequence = false): string {
        $year = (int)$requestDate->format('Y');
        $month = (int)$requestDate->format('n');
        $romanMonth = self::monthToRoman($month);

        $nextSequence = self::getNextPpcSequence($db, $year);
        $ppcNo = "{$nextSequence}/PPC/{$romanMonth}/{$year}";

        if ($recordSequence) {
            self::recordPpcSequence($db, $ppcNo, $year, $nextSequence);
        }

        return $ppcNo;
    }
}
